src/subset_matrix.js

```diff
@@ -0,0 +1,115 @@
+import * as scran from "scran.js"; 
+import * as utils from "./utils/general.js";
+
+/**
+ * This step subsets datasets based on a list of indices to continue with the rest of the analysis
+ *
+ * Methods not documented here are not part of the stable API and should not be used by applications.
+ * @hideconstructor
+ */
+export class SubsetCellsState {
+    #qc
+    #parameters;
+    #cache;
+
+    constructor(qc, parameters = null, cache = null) {
+        if (!(qc instanceof qc_module.QualityControlState)) {
+            throw new Error("'qc' should be a State object from './quality_control.js'");
+        }
+        this.#qc = qc;
+
+        this.#parameters = (parameters === null ? {} : parameters);
+        this.#cache = (cache === null ? {} : cache);
+        this.changed = false;
+    }
+
+    free() {
+        utils.freeCache(this.#cache.matrix);
+    }
+
+    /***************************
+     ******** Getters **********
+     ***************************/
+
+    fetchSubsetMatrix() {
+        if (!("matrix" in this.#cache)) {
+            this.#raw_compute(this.#parameters.indices);
+        }
+        return this.#cache.matrix;
+    }
+
+    fetchIndices() {
+        return this.#parameters.indices ? this.#parameters.indices : [];
+    }
+
+
+    /***************************
+     ******** Compute **********
+     ***************************/
+
+     #raw_compute(indices) {
+        var mat = this.#qc.fetchFilteredMatrix();
+
+        if (!indices || (Array.isArray(indices) && indices.length == 0)) {
+            this.#cache.matrix = mat;
+        } else {
+            this.#cache.matrix = scran.subsetColumns(mat, indices);
```

Comment:
Leaks memory, you don't free the existing `this.#cache.matrix` before reassigning it.

HOWEVER! You can't free the existing matrix if you assigned `mat` to it. This is because it would be a reference to the object owned up by the upstream step; a step doesn't have the authority to wipe the content of another step.

This implies that the only approach is to set `this.#cache.matrix = null` if no subsetting is requested, and then have `fetchSubsetMatrix` dynamically switch between returning the upstream matrix or `this.#cache.matrix`.